File: LeetCode/148. Sort List/Solution.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */
public class Solution
{
    public ListNode SortList(ListNode head)
    {
        return Divide(head);
    }

    private ListNode Divide(ListNode head)
    {
        var slow = head;
        var fast = head;
        var prev = slow;
        while (fast != null && fast.next != null)
        {
            prev = slow;
            slow = slow.next;
            fast = fast.next.next;
        }

        if (slow == head) return head;
        prev.next = null;
        return Merge(Divide(head), Divide(slow));
    }

    private ListNode Merge(ListNode head1, ListNode head2)
    {
        var head = head1;
        if (head1.val <= head2.val)
        {
            head = head1;
            head1 = head1.next;
        }
        else
        {
            head = head2;
            head2 = head2.next;
        }

        var cur = head;

        while (head1 != null || head2 != null)
        {
            if (head1 != null && head2 != null)
            {
                if (head1.val <= head2.val)
                {
                    cur.next = head1;
                    head1 = head1.next;
                }
                else
                {
                    cur.next = head2;
                    head2 = head2.next;
                }
            }
            else if (head1 != null)
            {
                cur.next = head1;
                head1 = head1.next;
            }
            else
            {
                cur.next = head2;
                head2 = head2.next;
            }

            cur = cur.next;
        }

        return head;
    }
}

File: LeetCode/148. Sort List/Solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def merge(self, head1: Optional[ListNode], head2: Optional[ListNode]) -> Optional[ListNode]:
        dummy = ListNode()
        cur = dummy
        while(head1 is not None or head2 is not None):
            if(head1 is not None and head2 is not None):
                if(head1.val <= head2.val):
                    cur.next = head1
                    head1 = head1.next
                else:
                    cur.next = head2
                    head2 = head2.next
            elif head1 is not None:
                cur.next = head1
                head1 = head1.next 
            else:
                cur.next = head2
                head2 = head2.next
            cur = cur.next
        return dummy.next

    def divide(self, head: Optional[ListNode]) -> Optional[ListNode]:
        fast = head
        slow = head
        prev = head
        while(fast is not None and fast.next is not None):
            prev = slow
            slow = slow.next
            fast = fast.next.next
        if(slow == head): return head
        prev.next = None
        return self.merge(self.divide(head), self.divide(slow))

    def sortList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        return self.divide(head)
